sklearn/preprocessing/_discretization.py

```diff
@@ -276,23 +246,25 @@ def transform(self, X):
         check_is_fitted(self, ["bin_edges_"])
         X = self._validate_X_post_fit(X)
 
-        Xt = _transform_selected(X, self._transform, self.dtype,
-                                 self.transformed_features_, copy=True,
-                                 retain_order=True)
+        X = check_array(X, dtype=FLOAT_DTYPES)
```

Comment:
We use ``X = check_array(X, accept_sparse='csc', copy=True, dtype=FLOAT_DTYPES)`` in ``_transform_selected``.
Here, I remove ``accept_sparse`` because we don't support sparse input here. I remove ``copy`` because I don't find it useful. I can't remove ``dtype`` because the result will change.